src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+/*! Standard Portable Intermediate Representation (SPIR-V) backend !*/
+use super::layout::{Instruction, Module};
+use crate::FastHashMap;
+use spirv::*;
+
+trait LookupHelper<T> {
+    type Target;
+    fn lookup_id(&self, handle: &crate::Handle<T>) -> Option<Word>;
+    fn lookup_handle(&self, word: &Word) -> Option<crate::Handle<T>>;
+}
+
+impl<T> LookupHelper<T> for FastHashMap<Word, crate::Handle<T>> {
+    type Target = T;
+
+    fn lookup_id(&self, handle: &crate::Handle<T>) -> Option<Word> {
+        let mut word = None;
+        for (k, v) in self.iter() {
+            if v.eq(handle) {
+                word = Some(*k);
+                break;
+            }
+        }
+        word
+    }
+
+    fn lookup_handle(&self, word: &u32) -> Option<crate::Handle<T>> {
+        let mut handle = None;
+        for (k, v) in self.iter() {
+            if k.eq(word) {
+                handle = Some(*v);
+                break;
+            }
+        }
+        handle
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct LookupFunctionType {
+    parameter_type_ids: Vec<Word>,
+    return_type_id: Word,
+}
+
+pub struct Parser {
+    module: Module,
+    id_count: u32,
+    capabilities: Vec<Capability>,
+    debugs: Vec<Instruction>,
+    annotations: Vec<Instruction>,
+    debug_enabled: bool,
+    void_type: Option<u32>,
+    lookup_type: FastHashMap<Word, crate::Handle<crate::Type>>,
+    lookup_function: FastHashMap<Word, crate::Handle<crate::Function>>,
+    lookup_function_type: FastHashMap<Word, LookupFunctionType>,
+    lookup_constant: FastHashMap<Word, crate::Handle<crate::Constant>>,
+    lookup_global_variable: FastHashMap<Word, crate::Handle<crate::GlobalVariable>>,
+    lookup_import: FastHashMap<Word, String>,
+    lookup_label: Vec<Word>,
+    current_label: Word,
+}
+
+impl Parser {
+    pub fn new(module: &crate::Module, debug_enabled: bool) -> Self {
+        Parser {
+            module: Module::new(module),
+            id_count: 0,
+            capabilities: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            debug_enabled,
+            void_type: None,
+            lookup_type: FastHashMap::default(),
+            lookup_function: FastHashMap::default(),
+            lookup_function_type: FastHashMap::default(),
+            lookup_constant: FastHashMap::default(),
+            lookup_global_variable: FastHashMap::default(),
+            lookup_import: FastHashMap::default(),
+            lookup_label: vec![],
+            current_label: 0,
+        }
+    }
+
+    fn generate_id(&mut self) -> Word {
+        self.id_count += 1;
+        self.id_count
+    }
+
+    fn bytes_to_words(&self, bytes: &[u8]) -> Vec<Word> {
+        let words: Vec<Word> = bytes
+            .chunks(4)
+            .map(|chars| match chars.len() {
+                4 => {
+                    (chars[3] as u32) << 24
```

Comment:
it looks like the whole `match` could be replaced by
```rust
chars.iter().rev().fold(0u32, |u, c| (u << 8) | c as u32)
```

Reply:
I am not yet sure, but it seems to work. It reverses all the characters in the iteration, then does `fold` which is something completely new for me and it produces one single output.

Reply:
Yeah, fold is a classy pure function, which can be used to implement a lot of different passes over data. As we are shifting the result to the left on each element, the first byte will end up in the high spot, hence I reversed the iterator to accomodate.